src/main/scala/scorex/core/ModifiersCache.scala

```diff
@@ -0,0 +1,139 @@
+package scorex.core
+
+
+import scorex.core.consensus.HistoryReader
+import scorex.core.utils.ScorexLogging
+import scorex.core.validation.RecoverableModifierError
+
+import scala.collection.mutable
+import scala.util.{Failure, Success}
+
+/**
+  * A cache which is storing persistent modifiers not applied to history yet.
+  * @tparam PMOD - type of a persistent node view modifier (or a family of modifiers).
+  */
+trait ModifiersCache[PMOD <: PersistentNodeViewModifier, H <: HistoryReader[PMOD, _]] {
+  require(maxSize >= 1)
+
+
+  protected val cache = mutable.Map[K, V]()
+
+  def size: Int = cache.size
+
+  /**
+    * How many elements are to be stored in the cache
+    */
+  def maxSize: Int
+
+  type K = mutable.WrappedArray[Byte]
+  type V = PMOD
+
+  /**
+    * Keys to simulate objects residing a cache. So if key is stored here,
+    * the membership check (contains()) shows that the key is in the cache,
+    * but the value corresponding to the key is not stored. The motivation
+    * to have this structure is to avoid repeatedly downloading modifiers
+    * which are unquestionably invalid.
+    */
+  protected val rememberedKeys = mutable.HashSet[K]()
```

Comment:
Should not it have limited size?

Reply:
@catena2w hard to say, in general case. Descendants are free to rewrite the default behavior, which is about to store (usually, 32) bytes for any invalid modifier, before a restart.